lib/eventstream_rpc.ts

```diff
@@ -0,0 +1,1153 @@
+     */
+    ClientStateError,
+
+    /**
+     * An error occurred ostensibly due to an underlying networking failure.
+     */
+    NetworkError,
+
+    /**
+     * An error that occurs when the underlying transport is shut down before an expected protocol event occurs.
+     */
+    InterruptionError,
+
+    /**
+     * Invalid data was passed into the RPC client.
+     */
+    ValidationError,
+
+    /**
+     * A formally-modeled error sent from server to client
+     */
+    ServiceError
+}
+
+/**
+ * @internal
+ */
+interface RpcErrorModel {
+    type: RpcErrorType;
+
+    description: string;
+
+    internalError?: CrtError;
+
+    serviceError?: any;
+};
+
+/**
+ * Wrapper type for all exceptions thrown by rpc clients and operations.  This includes rejected promises.
+ *
+ * The intention is for this data model to help users make better decisions in the presence of errors.  Not all errors
+ * are fatal/terminal, but JS doesn't really give a natural way to classify or conditionally react to general errors.
+ */
+export class RpcError extends Error {
+
+    /** The error's broad category */
+    readonly type: RpcErrorType;
+
+    /** Plain language description of the error */
+    readonly description: string;
+
+    /** Optional inner/triggering error that can contain additional context. */
+    readonly internalError?: CrtError;
+
+    /** Optional service-specific modelled error data */
+    readonly serviceError?: any;
+
+    constructor(model: RpcErrorModel) {
+        super(model.description);
+
+        this.type = model.type;
+        this.description = model.description;
+        if (model.internalError) {
+            this.internalError = model.internalError;
+        }
+        if (model.serviceError) {
+            this.serviceError = model.serviceError;
+        }
+    }
+}
+
+/**
+ * Wrapper for all data associated with an RPC client disconnection event
+ */
+export interface DisconnectionEvent {
+
+    /**
+     * Underlying reason for the disconnection
+     */
+    reason : CrtError;
+}
+
+/**
+ * Event listener type signature for listening to client disconnection events
+ */
+export type DisconnectionListener = (eventData?: DisconnectionEvent) => void;
+
+/**
+ * All data associated with the client successfully establishing an eventstream connection.
+ *
+ * Exists for future proofing at the moment.  Could eventually take connack properties, etc...
+ */
+export interface SuccessfulConnectionResult {
+
+}
+
+/**
+ * Configuration for the (potentially) asynchronous message transformation applied to the CONNECT message
+ * sent by the client once the underlying transport connection has been completed.
+ */
+export interface RpcMessageTransformationOptions {
+
+    /**
+     * (CONNECT) message to transform
+     */
+    message: eventstream.Message,
+
+    /**
+     * Optional controller that allows for cancellation of the asynchronous process.  The transformation implementation
+     * is responsible for respecting this.
```

Comment:
I think I know what you're saying here, but just to confirm: You mean that the transformation implementation is responsible for canceling the async process, correct? The canceling can only happen in the transformation?

Reply:
Yes